python/kvikio/kvikio/_lib/file_handle.pyx

```diff
@@ -0,0 +1,177 @@
+# Copyright (c) 2021-2024, NVIDIA CORPORATION. All rights reserved.
+# See file LICENSE for terms.
+
+# distutils: language = c++
+# cython: language_level=3
+
+import pathlib
+from typing import Optional
+
+from posix cimport fcntl
+
+from libc.stdint cimport uintptr_t
+from libcpp cimport bool
+from libcpp.string cimport string
+from libcpp.utility cimport move, pair
+
+from kvikio._lib.arr cimport parse_buffer_argument
+from kvikio._lib.future cimport (
+    IOFuture,
+    IOFutureStream,
+    _wrap_io_future,
+    _wrap_stream_future,
+    cpp_StreamFuture,
+    future,
+)
+
+from kvikio._lib import defaults
+
+
+cdef extern from "cuda.h":
+    ctypedef void* CUstream
```

Comment:
We should probably cimport the Cython stream type from cuda-python here.

Reply:
Using `from cuda.ccuda cimport CUstream`, I get:
```
building kvikio...
Using pip 24.2 from /home/mkristensen/apps/miniforge3/envs/kvikio-0809/lib/python3.11/site-packages/pip (python 3.11)
Obtaining file:///home/mkristensen/repos/kvikio/python/kvikio
  Running command Checking if build backend supports build_editable
  Checking if build backend supports build_editable ... done
  Running command Preparing editable metadata (pyproject.toml)
  *** scikit-build-core 0.10.1 using CMake 3.30.2 (metadata_editable)
  2024-08-12 15:43:46,567 - scikit_build_core - WARNING - No license files found, set wheel.license-files to [] to suppress this warning
  Preparing editable metadata (pyproject.toml) ... done
Building wheels for collected packages: kvikio
  Running command Building editable for kvikio (pyproject.toml)
  *** scikit-build-core 0.10.1 using CMake 3.30.2 (editable)
  2024-08-12 15:43:46,953 - scikit_build_core - WARNING - No license files found, set wheel.license-files to [] to suppress this warning
  *** Configuring CMake...
  2024-08-12 15:43:46,966 - scikit_build_core - WARNING - Unsupported CMAKE_ARGS ignored: -DCMAKE_BUILD_TYPE=Release
  2024-08-12 15:43:47,066 - scikit_build_core - WARNING - Unsupported CMAKE_ARGS ignored: -DCMAKE_BUILD_TYPE=Release
  loading initial cache file build/cp311-cp311-linux_x86_64/CMakeInit.txt
  -- KvikIO: Found cuFile Batch API: TRUE
  -- KvikIO: Found cuFile Stream API: TRUE
  -- Found nvcomp: /home/mkristensen/apps/miniforge3/envs/kvikio-0809/lib/cmake/nvcomp (found version 3.0.6)
  -- Configuring done (1.2s)
  -- Generating done (0.0s)
  -- Build files have been written to: /home/mkristensen/repos/kvikio/python/kvikio/build/cp311-cp311-linux_x86_64
  2024-08-12 15:43:48,280 - scikit_build_core - WARNING - Unsupported CMAKE_ARGS ignored: -DCMAKE_BUILD_TYPE=Release
  *** Building project with Ninja...
  [1/3] Transpiling file_handle.pyx to file_handle.cxx
  [2/3] Building CXX object kvikio/_lib/CMakeFiles/file_handle.dir/file_handle.cxx.o
  FAILED: kvikio/_lib/CMakeFiles/file_handle.dir/file_handle.cxx.o
  /home/mkristensen/apps/miniforge3/envs/kvikio-0809/bin/x86_64-conda-linux-gnu-c++ -DBS_THREAD_POOL_ENABLE_PAUSE=1 -DCUFILE_FOUND -DKVIKIO_CUDA_FOUND -DKVIKIO_CUFILE_BATCH_API_FOUND -DKVIKIO_CUFILE_FOUND -DKVIKIO_CUFILE_STREAM_API_FOUND -Dfile_handle_EXPORTS -isystem /home/mkristensen/apps/miniforge3/envs/kvikio-0809/include/python3.11 -isystem /home/mkristensen/apps/miniforge3/envs/kvikio-0809/include/gdeflate -fvisibility-inlines-hidden -fmessage-length=0 -march=nocona -mtune=haswell -ftree-vectorize -fPIC -fstack-protector-strong -fno-plt -O2 -ffunction-sections -pipe -isystem /home/mkristensen/apps/miniforge3/envs/kvikio-0809/include  -I/home/mkristensen/apps/miniforge3/envs/kvikio-0809/targets/x86_64-linux/include  -L/home/mkristensen/apps/miniforge3/envs/kvikio-0809/targets/x86_64-linux/lib -L/home/mkristensen/apps/miniforge3/envs/kvikio-0809/targets/x86_64-linux/lib/stubs -O3 -DNDEBUG -fPIC -pthread -MD -MT kvikio/_lib/CMakeFiles/file_handle.dir/file_handle.cxx.o -MF kvikio/_lib/CMakeFiles/file_handle.dir/file_handle.cxx.o.d -o kvikio/_lib/CMakeFiles/file_handle.dir/file_handle.cxx.o -c /home/mkristensen/repos/kvikio/python/kvikio/build/cp311-cp311-linux_x86_64/kvikio/_lib/file_handle.cxx
  /home/mkristensen/repos/kvikio/python/kvikio/build/cp311-cp311-linux_x86_64/kvikio/_lib/file_handle.cxx:1238:10: fatal error: cudaProfiler.h: No such file or directory
   1238 | #include "cudaProfiler.h"
        |          ^~~~~~~~~~~~~~~~
  compilation terminated.
  ninja: build stopped: subcommand failed.

  *** CMake build failed
```

Do we really need cudaProfiler?

Reply:
We want to use Cython types from cuda-python for consistency, but cuda-python exposes quite a few headers that may not be in the dependency list. Can you try adding a conda dependency on cuda-profiler-api (perhaps with a comment that it is needed by cuda-python)?

Reply:
installing `cuda-profiler-api` works, but their doesn't seems to be a `cuda-profiler-api<12` package?

Reply:
You will need that package only for building packages of CUDA 12. For CUDA 11 and CUDA 12 wheel builds, it is expected that the system provides all CUDA headers.

Reply:
Does this look correct? 
https://github.com/rapidsai/kvikio/pull/434/commits/d965e66a1e297e740a9a385e2cc7ba49a6adee1d

Reply:
Fails here: https://github.com/rapidsai/kvikio/actions/runs/10354623316/job/28660538187?pr=434#step:9:233

Maybe, this is something we should look at in a follow up PR?

Reply:
I have reverted the change, let's do it in another PR. I like to keep this PR to refactoring

Reply:
Let's raise an issue to track. @bdice would you be able to help with this?

Reply:
No, I don’t think I can take this on. It is a low priority.

Reply:
For clarity just meant raising an issue. That said, if we don't think it is worth an issue, it is likely not worth doing either